File: android/app/src/main/java/gg/yappy/app/ui/chat/LocationShare.kt
package gg.yappy.app.ui.chat

import android.content.Intent
import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.LocationOn
import androidx.compose.material.icons.rounded.Map
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import gg.yappy.app.data.LiveLocation
import gg.yappy.app.data.LocationPayload
import gg.yappy.app.ui.theme.neuColors
import gg.yappy.app.ui.components.NeuButton
import gg.yappy.app.ui.util.Locator
import gg.yappy.app.ui.util.clockTime
import java.time.Instant

/**
 * Choosing what to share.
 *
 * Two things, and the second is why this is a sheet rather than a button:
 * sending a pin is harmless and instant, while sharing live means agreeing to
 * be followed for a while. Those should not be one tap apart with no chance to
 * think about it.
 */
@Composable
fun LocationShareSheet(onShare: (Long?) -> Unit) {
    val colors = neuColors
    val context = LocalContext.current
    var granted by remember { mutableStateOf(Locator.granted(context)) }
    var working by remember { mutableStateOf(false) }

    val permission = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions(),
    ) { result -> granted = result.values.any { it } }

    LaunchedEffect(Unit) {
        if (!granted) {
            permission.launch(
                arrayOf(
                    android.Manifest.permission.ACCESS_FINE_LOCATION,
                    android.Manifest.permission.ACCESS_COARSE_LOCATION,
                ),
            )
        }
    }

    Column(
        Modifier.fillMaxWidth().padding(22.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Text("Share location", style = MaterialTheme.typography.titleSmall, color = colors.textPrimary)

        if (!granted) {
            Text(
                "yappy needs permission to use your location.",
                style = MaterialTheme.typography.bodyMedium,
                color = colors.textSecondary,
            )
            NeuButton(onClick = {
                permission.launch(
                    arrayOf(
                        android.Manifest.permission.ACCESS_FINE_LOCATION,
                        android.Manifest.permission.ACCESS_COARSE_LOCATION,
                    ),
                )
            }) { Text("Allow location", style = MaterialTheme.typography.labelMedium) }
            return@Column
        }

        NeuButton(
            onClick = { if (!working) { working = true; onShare(null) } },
            modifier = Modifier.fillMaxWidth(),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(Icons.Rounded.LocationOn, contentDescription = null, modifier = Modifier.size(18.dp))
                Text("  Send my current location", style = MaterialTheme.typography.labelMedium)
            }
        }

        Text("OR SHARE LIVE", style = MaterialTheme.typography.labelSmall, color = colors.textTertiary)
        Text(
            "Everyone here sees you move until it ends. You can stop at any time.",
            style = MaterialTheme.typography.labelSmall,
            color = colors.textTertiary,
        )

        // The same three the server allows, and the same reasoning: each is a
        // length somebody can hold in their head as "until roughly when".
        listOf("15 minutes" to 900L, "1 hour" to 3_600L, "8 hours" to 28_800L).forEach { (label, seconds) ->
            NeuButton(
                onClick = { if (!working) { working = true; onShare(seconds) } },
                modifier = Modifier.fillMaxWidth(),
            ) {
                Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    Icon(Icons.Rounded.Map, contentDescription = null, modifier = Modifier.size(18.dp))
                    Text("  $label", style = MaterialTheme.typography.labelMedium)
                }
            }
        }
    }
}

/**
 * A shared place, in the timeline.
 *
 * No map drawn here, and that is a decision rather than a gap. Android has no
 * free map surface: the Google SDK needs an API key and Play Services, and
 * scraping somebody's public tile server is not something to ship. So the card
 * says what it knows and hands off to whichever maps app the phone already has
 * — which is where people end up anyway.
 */
@Composable
fun LocationCard(
    payload: LocationPayload,
    live: LiveLocation?,
    isMine: Boolean,
    onStop: () -> Unit,
) {
    val colors = neuColors
    val context = LocalContext.current

    val latitude = live?.latitude ?: payload.latitude
    val longitude = live?.longitude ?: payload.longitude

    /** Still moving: not ended, and not past its end time. */
    val isLive = live != null && live.endedAt == null &&
        runCatching { Instant.parse(live.expiresAt).isAfter(Instant.now()) }.getOrDefault(false)

    Column(
        Modifier
            .width(240.dp)
            .clip(RoundedCornerShape(18.dp))
            .background(colors.veil)
            .clickable {
                // `geo:` with a label, so the pin lands named rather than as a
                // bare coordinate somebody has to squint at.
                val label = Uri.encode(payload.name ?: "Shared location")
                val uri = Uri.parse("geo:$latitude,$longitude?q=$latitude,$longitude($label)")
                runCatching { context.startActivity(Intent(Intent.ACTION_VIEW, uri)) }
            }
            .padding(14.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Box(
                Modifier
                    .size(34.dp)
                    .clip(CircleShape)
                    .background(if (isLive) colors.success.copy(alpha = 0.18f) else colors.accent.copy(alpha = 0.18f)),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    Icons.Rounded.LocationOn,
                    contentDescription = null,
                    tint = if (isLive) colors.success else colors.accent,
                    modifier = Modifier.size(19.dp),
                )
            }
            Column(Modifier.padding(start = 10.dp)) {
                Text(
                    payload.name?.takeIf { it.isNotBlank() } ?: "Location",
                    style = MaterialTheme.typography.labelMedium,
                    color = colors.textPrimary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Text(
                    // Five decimals is about a metre. More is noise, and fewer
                    // puts you on the wrong side of the street.
                    "%.5f, %.5f".format(latitude, longitude),
                    style = MaterialTheme.typography.labelSmall,
                    color = colors.textTertiary,
                    maxLines = 1,
                )
            }
        }

        when {
            isLive && live != null -> {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Box(Modifier.size(7.dp).clip(CircleShape).background(colors.success))
                    Text(
                        "  Live until ${clockTime(live.expiresAt)}",
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.success,
                    )
                }
                if (isMine) {
                    Text(
                        "Stop sharing",
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.danger,
                        modifier = Modifier.clickable(onClick = onStop),
                    )
                }
            }
            // A share that has finished. Saying so is what stops the last known
            // point being read as where somebody is now.
            payload.liveUntil != null -> Text(
                "Live location ended",
                style = MaterialTheme.typography.labelSmall,
                color = colors.textTertiary,
            )
            else -> Text("Tap to open in Maps", style = MaterialTheme.typography.labelSmall, color = colors.textTertiary)
        }
    }
}

File: ios/Yappy/Features/Chat/LocationShare.swift
import CoreLocation
import MapKit
import SwiftUI

/// Choosing what to share.
///
/// Two things, and the second one is the reason this is a sheet rather than a
/// button: sending a pin is harmless and instant, while sharing live means
/// agreeing to be followed for a while, and those should not be one tap apart
/// with no chance to think.
///
/// The duration is the whole decision, so the options are the ones people
/// actually mean — "I am nearly there", "I am on my way", "we are out for the
/// day" — and every one of them is a length someone can hold in their head.
struct LocationShareSheet: View {
    @Environment(\.neu) private var colors
    @Environment(\.dismiss) private var dismiss
    @StateObject private var locator = Locator.shared

    /// nil duration = a pin, right now. Non-nil = live for that long.
    let onShare: (TimeInterval?) -> Void

    @State private var fix: CLLocation?
    @State private var working = false

    private static let durations: [(String, TimeInterval)] = [
        ("15 minutes", 900),
        ("1 hour", 3_600),
        ("8 hours", 28_800),
    ]

    var body: some View {
        VStack(alignment: .leading, spacing: 18) {
            Text("Share location")
                .font(YappyFont.titleSmall)
                .foregroundStyle(colors.text)

            if locator.denied {
                denied
            } else if !locator.authorised {
                asking
            } else {
                preview
                pin
                live
            }
        }
        .padding(22)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(colors.surface)
        .task {
            locator.requestAuthorisation()
            fix = await locator.current()
        }
    }

    // ── States ───────────────────────────────────────────────────────────────

    private var asking: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("yappy needs permission to use your location.")
                .font(YappyFont.body)
                .foregroundStyle(colors.textSecondary)
            NeuButton(action: { locator.requestAuthorisation() }) {
                Text("Allow location").font(YappyFont.labelMedium)
            }
        }
    }

    private var denied: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Location is turned off for yappy.")
                .font(YappyFont.body)
                .foregroundStyle(colors.textSecondary)
            Text("Turn it on in Settings › Privacy › Location Services.")
                .font(YappyFont.labelSmall)
                .foregroundStyle(colors.textTertiary)
        }
    }

    /// The map is here so nobody sends a pin without seeing what it is. A fix
    /// that has not arrived shows as a spinner rather than a wrong place.
    @ViewBuilder
    private var preview: some View {
        if let fix {
            Map(initialPosition: .region(MKCoordinateRegion(
                center: fix.coordinate,
                span: MKCoordinateSpan(latitudeDelta: 0.004, longitudeDelta: 0.004)
            ))) {
                Marker("You", coordinate: fix.coordinate)
                    .tint(colors.accent)
            }
            .frame(height: 170)
            .clipShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
            .allowsHitTesting(false)
        } else {
            RoundedRectangle(cornerRadius: 18, style: .continuous)
                .fill(colors.veil)
                .frame(height: 170)
                .overlay(ProgressView())
        }
    }

    private var pin: some View {
        NeuButton(action: { share(nil) }) {
            HStack(spacing: 10) {
                Image(systemName: "mappin.circle.fill")
                Text("Send this location").font(YappyFont.labelMedium)
            }
        }
        .disabled(fix == nil || working)
        .opacity(fix == nil ? 0.5 : 1)
    }

    private var live: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("OR SHARE LIVE")
                .font(YappyFont.labelSmall)
                .foregroundStyle(colors.textTertiary)
            Text("Everyone here sees you move until it ends. You can stop at any time.")
                .font(YappyFont.labelSmall)
                .foregroundStyle(colors.textTertiary)
                .fixedSize(horizontal: false, vertical: true)

            ForEach(Self.durations, id: \.0) { label, seconds in
                NeuButton(action: { share(seconds) }) {
                    HStack {
                        Image(systemName: "dot.radiowaves.left.and.right")
                        Text(label).font(YappyFont.labelMedium)
                        Spacer()
                    }
                }
                .disabled(fix == nil || working)
                .opacity(fix == nil ? 0.5 : 1)
            }
        }
    }

    private func share(_ duration: TimeInterval?) {
        guard fix != nil, !working else { return }
        working = true
        onShare(duration)
        dismiss()
    }
}

// ── The bubble ───────────────────────────────────────────────────────────────

/// A shared place, in the timeline.
///
/// A real map rather than a picture of one: MapKit costs nothing on iOS, needs
/// no key, and a static image would still have to be fetched from somebody's
/// tile server. The dot moves as pings arrive, so a live share is the same view
/// with different coordinates rather than a second kind of card.
struct LocationBubble: View {
    @Environment(\.neu) private var colors

    let payload: LocationPayload
    /// The current point, when this share is still moving. Nil means the map
    /// shows where it started, which is all a finished share has to say.
    let live: LiveLocation?
    let isMine: Bool
    let onStop: () -> Void
    let onOpen: () -> Void

    private var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(
            latitude: live?.latitude ?? payload.latitude,
            longitude: live?.longitude ?? payload.longitude
        )
    }

    /// A live share that has not ended and has not run out.
    private var isLive: Bool {
        guard let live, live.endedAt == nil else { return false }
        guard let until = YappyTime.parse(live.expiresAt) else { return false }
        return until > Date()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            map
            footer
        }
        .frame(width: 250)
        .background(colors.veil, in: RoundedRectangle(cornerRadius: 20, style: .continuous))
        .contentShape(Rectangle())
        .onTapGesture(perform: onOpen)
    }

    private var map: some View {
        Map(position: .constant(.region(MKCoordinateRegion(
            center: coordinate,
            span: MKCoordinateSpan(latitudeDelta: 0.004, longitudeDelta: 0.004)
        )))) {
            Marker(payload.name ?? "Here", coordinate: coordinate)
                .tint(isLive ? colors.success : colors.accent)
        }
        .frame(height: 140)
        .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
        // The map is a picture here. Panning it inside a scrolling timeline
        // fights the list for every gesture, and the tap opens Maps anyway.
        .allowsHitTesting(false)
    }

    private var footer: some View {
        VStack(alignment: .leading, spacing: 6) {
            if let name = payload.name, !name.isEmpty {
                Text(name)
                    .font(YappyFont.labelMedium)
                    .foregroundStyle(colors.text)
                    .lineLimit(1)
            }

            if isLive, let live {
                HStack(spacing: 6) {
                    Circle()
                        .fill(colors.success)
                        .frame(width: 7, height: 7)
                    Text("Live until \(YappyTime.clockTime(live.expiresAt))")
                        .font(YappyFont.labelSmall)
                        .foregroundStyle(colors.success)
                }
                if isMine {
                    Button(action: onStop) {
                        Text("Stop sharing")
                            .font(YappyFont.labelSmall)
                            .foregroundStyle(colors.danger)
                    }
                    .buttonStyle(.plain)
                }
            } else if payload.liveUntil != nil {
                // A share that has finished. Saying so is what stops the last
                // known point from being read as where somebody is now.
                Text("Live location ended")
                    .font(YappyFont.labelSmall)
                    .foregroundStyle(colors.textTertiary)
            } else {
                Text("Location")
                    .font(YappyFont.labelSmall)
                    .foregroundStyle(colors.textTertiary)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 14)
        .padding(.vertical, 10)
    }
}

/// Hand off to Maps. Nothing in the app is a better map than the one the phone
/// already has, and "open in Maps" is what people do with a pin anyway.
func openInMaps(_ payload: LocationPayload, current: LiveLocation?) {
    let coordinate = CLLocationCoordinate2D(
        latitude: current?.latitude ?? payload.latitude,
        longitude: current?.longitude ?? payload.longitude
    )
    let item = MKMapItem(placemark: MKPlacemark(coordinate: coordinate))
    item.name = payload.name ?? "Shared location"
    item.openInMaps()
}
